# Individual Assignment Week 3 (Midterm take-home exam) | Analysing 40 years of Gary Numan

### Author: *George Christian Cotea*

<tr>
    <td> <img src="https://upload.wikimedia.org/wikipedia/en/2/2e/ThePleasurePrinciple1.jpg" width="250" height ="250"/> </td>

</tr>


 

The British singer Gary Numan is often regarded as a pioneer of electronic music, debuting as a solo artist in 1979 with [The Pleasure Principle](https://open.spotify.com/album/6wHjdKs7VVPVcqaHRzwqJt), an album that reached the top of the UK Albums Chart. The song "Cars" from this album became a worldwide hit. The album is notable for relying heavily on synthesizers, with no electric guitar, yet having a rock structure. Befitting of this futuristic-at-the-time sound, the lyrics cover science fiction themes.

Over time, Gary Numan has changed his style and sound several times, but he is still releasing albums which are well regarded and are considered to have a contemporary sound. It would be interesting to see how much an artist's lyrics change over such a long timespan of about 40 years. In this assignment, we will explore the lyrics of the 1979 album The Pleasure Principle and of his 2017 album, [Savage](https://open.spotify.com/album/2d98ICB4Uin8lszUnnCwxo), which again covers science fiction themes.

Remember that this is an individual assignment. Please have a look at the requirements of this assignment on Canvas. In particular, if you take code from external resources for this assignment, then give a citation in the forum of a URL, and describe what it is doing. The UvA rules for plagiarism apply.

Note that unlike the collaborative assignments, this assignment is not only graded for correctness, but also for general criteria such as presentation (including documentation and proper use of text cells), structure and problem solving, and ingenuity/originality. Have a look at the grading rubric on Canvas for more information. Don't forget the questions that don't ask for code.

Also remember that you can download the .ipnyb file from the File menu to submit on Canvas.

All the very best!

## Data

As our data, we will use the lyrics of 5 songs from the 1979 album and 5 songs from the 2017 album.

Normally, I would offer these lyrics as files, but we have not learned to work with files yet. So, they are pasted below in the form of two Python dictionaries, one for the 1979 album and one for the 2017 album.

Feel free to have a look at this data before proceeding.
(and giving the album a listen perhaps during analysis :))

FYI: This data was acquired from [Genius](https://genius.com) (Genius Lyrics) using their API, something that you could do too using the lyricsgenius package for Python!

In [1]:
#Dictionary for the lyrics of 5 songs from the 1979 album
numan_lyrics_1979 = {
"Metal" : """We're in the building where they make us grow
And I'm frightened by the liquid engineers
Like you
My Mallory heart is sure to fail
I could crawl around the floor just like I'm real
Like you
The sound of metal
I want to be you
I could learn to be a man
Like you
Plug me in
And turn me on
Oh, everything is moving
I need my treatment, it's tomorrow they send me
Singing, "I am an American"
Do you?
Picture this, if I could make the change
I'd love to pull the wires from the wall
Did you?
And who are you? And how can I try?
Here inside, I like metal
Don't you?
All I know
Is no one dies
I'm still confusing love with need""",
"M.E." : """And M.E., I eat dust
We're all so run down
I'd call it my death
But I'll only fade away
And I hate to fade alone
Now there's only M.E

We were so sure
We were so wrong
Now it's over
But there's no one left to see
And there's no one left to die
There's only M.E

Why should I care?
Why should I try?
Oh no, oh no
I turned off the pain
Like I turned off you all
Now there's only M.E""",
"Conversation" : """Oh, it's so easy
When parts take over
My conversation
Is nothing more than lies

You're just the viewer
So cold and distant
I've no intentions
Of saying "I love you"
My conversation
My conversation
My conversation
My conversation

And we are not gods
And we are not men
And we are not making claims
We are only boys

And you are not strong
And you are not force
And you are not regular
You are just wrong

There are no faces
This is my complex
You are my picture
I call you "mirrors"

These are not my tears
Not my reflection
Am I a photo?
I can't remember
My conversation
My conversation
My conversation
My conversation
My conversation
My conversation
My conversation
My conversation
My conversation
My conversation
My conversation""",
"Cars" : """[Verse]
Here in my car
I feel safest of all
I can lock all my doors
It's the only way to live
In cars
Here in my car
I can only receive
I can listen to you
It keeps me stable for days
In cars
[Verse]
Here in my car
Where the image breaks down
Will you visit me please
If I open my door
In cars?
Here in my car
I know I've started to think
About leaving tonight
Although nothing seems right
In cars""",
"Engineers" : """All that we are
Is all that we need to be
All that we know
Is you and machinery
We're engineers
We are your heartbeat
We are your night life
We are your low-line
We keep you alive for now
We're engineers
We are your voice
We are your blood flow
We are your eyes
We're all you need to know
We're engineers
All that we are
Is all that you'd love to be
All that we know
Is hate and machinery
We're engineers"""}

#Dictionary for the lyrics of 5 songs from the 2017 album
numan_lyrics_2017 = {
"Ghost Nation" : """[Verse]
We live in a windswept hell
Where dust and death are neighbours
We hide in a perfect storm
Not even God remembers

[Verse]
We ran from the wars of man
From everything you are and would be
We are invisible
We are the dead Ghost Nation

[Chorus]
When the sky came down
When the sun went dark
When the night was torn
And we heard God laugh
When The Righteous came
And they cleansed our sins
Have you ever seen death
With your own eyes?

[Verse]
We live in The Great Unfound
Where thought and ghost still wander
We drift on drifting sand
Where horrors of our past still linger

[Verse]
We hide from the world of man
From the cruelty that defines you
We are the wind that cries
We are the dead Ghost Nation

[Chorus]
When the sky came down
When the sun went dark
When the night was torn
And we heard God laugh
When the Righteous came
And they cleansed our sins
Have you ever seen God
With your own eyes?

[Bridge]
We are embers in the the flame
We are whispers in God's name
We're the heat that burns your skin
We’re the cold that lets pain in
We’re the wind that blinds your eyes
We’re the shame that haunts your lies
We’re the nightmare when you sleep
We’re the guilt that all men keep

[Chorus]
When the sky came down
When the sun went dark
When the night was torn
And we heard God laugh
When the Righteous came
And they cleansed our sins
Have you ever seen me
With your own eyes?""",
"My Name Is Ruin" : """When they called me broken, I knew
When they called me evil, I knew
When they called me Ruin, I knew
I would always find my way to you

When I begged forgiveness, they knew
When I begged for mercy, they knew
When I begged for nothing, they knew
I would always find my way to you

My name is Ruin, my name is vengeance
My name is no-one, no-one is calling
My name is Ruin, my name is heartbreak
My name is lonely, my sorrow's a darkness
My name is Ruin, my name is evil
My name's a war song, I sing you a new one
My name is ruin, my name is broken
My name is shameless, I'll tear your world open

When I called you poison, you knew
When I called you shameful, you knew
When I called you a liar, you knew
I would always find my way to you

I'll show you ruin, I'll show you vengeance
I'll show you no one, and no one is calling
I'll show you ruin, I'll show you heartbreak
I'll show you loving, and sorrow and darkness
I'll show you ruin, I'll show you evil
I'll sing you a war song, I'll sing you a new one
I'll show you ruin, I'll show you broken
I'll show you shameless, I'll tear your world open""",
"The End of Things" : """Is that a voice calling me softly?
Nothing in here is quite how it seems
Is that a whisper's touch or my fear?
Am I awake or lost in old dreams?
Everything's so cold, the air is so still
There should be nothing here to see
Is this a warning, like in the Black Book?
There can be no one else but Me
I see a darker shade of darkness
I hear a shadow hide in the lie
I feel a suffocating tension
Within a terrible desire
Everything is so cold, the air is so still
There should be nothing here to fear
But I hear breathing, a child's laughter
Can any part of this be real?

Everything I work for
Everything I long for is always just too far
Everything I hope for never comes to be
Everything I bleed for burns a scar on me
Everything I fight for leaves a bitter taste
Everything I cry for laughs into my face
Everything I scream for barely knows my name
Everything I'd die for will die just the same
In here, with me

Is this the truth of your forgiveness?
Are you the stranger stood at my door?
Are you the end of things come calling?
Are you the answer that I've wished for?
Everything's so cold, the air is so still
And there is nothing here but me
If I belong here, and this is mercy
Then there is no place I'd rather be

Everything I worked for
Everything I longed for was always just too far
Everything I hoped for never came to be
Everything I bled for burned a scar on me
Everything I fought for left a bitter taste
Everything I cried for laughed into my face
Everything I screamed for barely knew my name
Everything I died for died just the same
In here, with me""",
"When the World Comes Apart" : """I'm not a hero's story
I'm not a traitor's lie
I'm not everything they said
I'm not here to save you
I'm not your confessor
My sins alone can wake the dead

But do you understand
Nothing can make it better
Nothing can make it go away
And do you understand
That we are lost to ruin
And this will be your end of days

And when the sun fell down
And when the moon failed to rise
And when the world came apart
Where were you?
Were you with me?
When my light burns out
And when my fire is cold
And when my breath is the wind
Where will you be, dear God?

I have no forgiveness
I have a heart of stone
I am a lesson born of greed
I am torn and twisted
I am a broken bone
I am what you made of me

But do you understand
That none of this will matter
Nothing can take your pain away
And do you understand
That everything is shattered
And so the whole world ends today

And when the sun fell down
And when the moon failed to rise
And when the world came apart
Where were you?
Were you with me?
When my light burns out
And when my fire is cold
And when my breath is the wind
Where will you be, dear God?
And when the sun fell down
And when the moon failed to rise
And when the world came apart
Where were you?
Were you with me?
When my light burns out
And when my fire is cold
And when my breath is the wind
I will find you, dear God""",
"If I Said" : """If I said I was a painter would you love and comfort me?
If I said I was a soldier would you cry a tear for me?
If I said I was a poet would you share your hopes with me?
If I said I was a dreamer would you share your life with me?
If I asked you for forgiveness would you smile and wonder why?
Would you say that you believed me or would it all fall down?
Would it all fall down?
Is it over; is it over now?
If I said I was a liar would you lose your faith in me?
If I said I was mistaken would you share the blame with me?
If I said I was a memory would you still remember me?
If I said I was a warning would you shy away from me?
If I asked you for forgiveness would you turn your face away?
Would you say I broke your heart and would it all fall down?
Would it all fall down?
Is it over; is it over now?
When a new tear falls
When I hear you cry softly
I would give the world to change the past
And when you walked away
And I saw what I was losing
Only me to blame, I wondered why, why, why
So it all comes back to haunt me
Like a song that sings your name
And it burns here to remind me
That it all comes back as pain
When I asked you for forgiveness and the words drifted away
I watched them float in empty silence as it all fell down
As it all fell down
It was over
It was over... then"""
}

## Exercises


We would like to use Python code to look through the sampled lyrics of the two time periods and be able to understand their characteristics. 

Make your results look as pretty as possible using the things you have learned (remember f-strings), and feel free to use spacing or enumeration.

1. Write a **function** that reads through the songs in a single dictionary and returns all the song titles in that dictionary.

    *   Use this function to read through both album dictionaries and print the titles from each.


In [2]:
# In the dictionaries provided, the song titles are the keys, I will define a function that reads the keys and returns the song titles as a list.
def dicTitles(albumLyrics):
    '''This function reads the keys of a dictionary containing the song titles as keys and the lyrics as values and returns the song titles as a list.
    Input: albumLyrics (dictionary)
    Output: titles (list)'''
    titles = []
    for key in albumLyrics: # this is a for loop that appends the titles list with each key (song title) in the dictionary
        titles.append(key)
    return titles
#Return the titles of the songs in the two albums:
print(f"The titles of the 1979 album are: {dicTitles(numan_lyrics_1979)}")
print(f"The titles of the 2017 album are: {dicTitles(numan_lyrics_2017)}")

The titles of the 1979 album are: ['Metal', 'M.E.', 'Conversation', 'Cars', 'Engineers']
The titles of the 2017 album are: ['Ghost Nation', 'My Name Is Ruin', 'The End of Things', 'When the World Comes Apart', 'If I Said']


2. Perform text normalization on the lyrics. Text normalisation should include removal of

    *   neccessary punctuations
    *   lowercasing the lyrics
    *   remove meta-information that is not part of the actual lyrics
    
    <br/>
    You can also do any further normalization or pre-processing that you think would make the analysis in questions 3 and 4 better. Keywords for researching this topic: "text normalization", "stop words", "tokenization", "text preprocessing". However keep in mind that these are lyrics which is a bit different than other types of text.
    
   * Provide arguments for any decisions you make regarding further processing in your notebook, even if you decide not to do any.

For this function I will only do text normalization as I think that the lyrics are already quite clean and do not need any further processing. I will remove punctuation, lowercase the lyrics and remove meta-information that is not part of the actual lyrics. I will also remove the song titles from the lyrics as they are not part of the actual lyrics.

Removing stop words and tokenization are not necessary for this analysis as we are not looking for specific words, but rather the overall characteristics of the lyrics. If we were looking for specific words, then removing stop words and tokenization would be necessary.

*insert text preprocessing bit later here*

In [3]:
# Import the necessary libraries
import re
import string
# Defining a function that normalizes the lyrics in a dictionary containing the song titles as keys and the lyrics as values.
def normalizeLyrics(dictionary):
    '''This function normalizes the lyrics in a dictionary containing the song titles as keys and the lyrics as values and stores them in a new dictionary for further processing in exercise 3 and 4.
    Input: dictionary (dictionary)
    Output: normalized dictionary (dictionary)'''

    # Create a new dictionary to store the normalized values
    normalized_dict = {}

    # Loop through the input dictionary
    for x, lyrics in dictionary.items():
         # Remove meta-information that is not part of the actual lyrics
        lyrics = re.sub(r'\[.*?\]', '', lyrics) #removes the [Verse] and [Chorus] tags
        lyrics = re.sub(r"\(.*?\)", '', lyrics) #removes the (Instrumental) tags from the 2017 album and the (x2) tags from the 1979 album
        # Remove punctuation
        lyrics = lyrics.translate(str.maketrans('', '', string.punctuation))
        # Lowercase the text
        lyrics = lyrics.lower() #lowercasing the lyrics helps count the same word in different cases as the same word.
        # Remove extra whitespace
        lyrics = re.sub('\s+', ' ', lyrics).strip() #removes extra whitespace that was used as formatting for the lyrics

        # Add the normalized lyrics to the new dictionary
        normalized_dict[x] = lyrics

    return normalized_dict

print(normalizeLyrics(numan_lyrics_1979))
normalized_1979 = normalizeLyrics(numan_lyrics_1979)
print(normalizeLyrics(numan_lyrics_2017))
normalized_2017 = normalizeLyrics(numan_lyrics_2017)


{'Metal': 'were in the building where they make us grow and im frightened by the liquid engineers like you my mallory heart is sure to fail i could crawl around the floor just like im real like you the sound of metal i want to be you i could learn to be a man like you plug me in and turn me on oh everything is moving i need my treatment its tomorrow they send me singing i am an american do you picture this if i could make the change id love to pull the wires from the wall did you and who are you and how can i try here inside i like metal dont you all i know is no one dies im still confusing love with need', 'M.E.': 'and me i eat dust were all so run down id call it my death but ill only fade away and i hate to fade alone now theres only me we were so sure we were so wrong now its over but theres no one left to see and theres no one left to die theres only me why should i care why should i try oh no oh no i turned off the pain like i turned off you all now theres only me', 'Conversation

3. Write a function that returns some statistics about the two albums' song lyrics, and run this function for both albums:

    * Most frequent words
    * Type to token ratio (unique words / total words)
    * Average word length
    * Longest and shortests songs (by lyrics)
    * What are the songs with the largest vocabulary and smallest vocabulary?
    
    <br/>
    Print these results to your notebook for both albums in a nice looking way. Then, also show these same statistics for just the song 'Cars' from the 1979 album, which was Gary Numan's most famous song.
    
   * Write down your interpretation of these results in this notebook.  
   * In which of the two time periods was Gary Numan more verbose? Back it with some evidence.
   * Electronic music is sometimes said to make more use of repetition than other forms of music. In which of the two time periods did Gary Numan make more use of lyrical repetition? You can either argue your case based on the numbers you were asked to calculate, or you can come up with your own definition of 'repetitiveness' and calculate it with Python code.

In [4]:
# Making a list of all the normalized lyrics from the 1979 album
lyrics_1979 = list(normalizeLyrics(numan_lyrics_1979).values())
#Making a list of all the normalized lyrics from the 2017 album
lyrics_2017 = list(normalizeLyrics(numan_lyrics_2017).values())
#Creating a function that returns the most frequent words in a list of lyrics
from collections import Counter
def mostFrequentWords(lyrics):
    '''This function returns the most frequent words in a list of lyrics.
    Input: lyrics (list)
    Output: most frequent words (list)'''
    words = []
    for lyric in lyrics: #this is a for loop that appends the words list with each word in the lyrics
        words.extend(lyric.split())
    return Counter(words).most_common(15) #I chose to limit the return to the 15 most frequent words for cleanliness
    #return Counter(words).most_common() #This variant of code has no limits and returns all the words
#Converting the list of touples provided by the counter to a dictionary for later use
mostFrequentWordsDict1979 = dict(mostFrequentWords(lyrics_1979))

#Creating a function that returns the type to token ratio in a list of lyrics
def typeToTokenRatio(lyrics):
    '''This function returns the type to token ratio in a list of lyrics.
    Input: lyrics (list)
    Output: type to token ratio (float)'''
    words = []
    for lyric in lyrics: #this is a for loop that appends the words list with each word in the lyrics
        words.extend(lyric.split())
    return len(set(words))/len(words)
#Creating a function that returns the average word length in a list of lyrics
def averageWordLength(lyrics):
    '''This function returns the average word length in a list of lyrics.
    Input: lyrics (list)
    Output: average word length (float)'''
    words = []
    for lyric in lyrics: #this is a for loop that appends the words list with each word in the lyrics
        words.extend(lyric.split())
    return sum(len(word) for word in words)/len(words)
#Creating a function that returns the longest and shortest songs in a list of lyrics
def longestAndShortestSongs(normalized_dictionary):
    '''This function returns the longest and shortest songs in a list of lyrics.
    Input: normalized_dictionary (dictionary)
    Output: longest and shortest songs (list)
    '''
    max_song_len = 0
    min_song_len = 9999999
    max_song_name = ''
    min_song_name = ''
    for song in normalized_dictionary:
        l = len(normalized_dictionary[song])
        if l > max_song_len:
            max_song_len = l
            max_song_name = song
        if l < min_song_len:
            min_song_len = l
            min_song_name = song
    return [max_song_name, min_song_name]
#Creating a function that returns the songs with the largest and smallest vocabulary in a list of lyrics
def largestAndSmallestVocabulary(normalized_dictionary):
    '''This function returns the songs with the largest and smallest vocabulary in a list of lyrics.
    Input: lyrics (list)
    Output: songs with the largest and smallest vocabulary (list)'''
    max_song_len = 0
    min_song_len = 9999999 #I chose a ridiculously large number to make sure that the first song will be the smallest
    max_song_name = ''
    min_song_name = ''
    for song in normalized_dictionary:
        l = len(set(normalized_dictionary[song]))
        if l > max_song_len:
            max_song_len = l
            max_song_name = song
        if l < min_song_len:
            min_song_len = l
            min_song_name = song
    return [max_song_name, min_song_name]
#Introducing intentional waiting for UX purposes
import time
print('Calculating statistics for the 1979 album...')
time.sleep(1)
print('Conjuring up the statistics...\n')
time.sleep(2)

#Checking the most frequent words in the 1979 album
mostFrequentWordsDict1979 = dict(mostFrequentWords(lyrics_1979)) #Converting the list of touples provided by the counter to a dictionary for later use
print(f"The most frequent words in the 1979 album are: {mostFrequentWordsDict1979}\n")
time.sleep(1)
#Checking the type to token ratio in the 1979 album
print(f"The type to token ratio in the 1979 album is: {typeToTokenRatio(lyrics_1979)}\n")
time.sleep(1)
#Checking the average word length in the 1979 album
print(f"The average word length in the 1979 album is: {averageWordLength(lyrics_1979)} words.\n")
time.sleep(1)
#Checking the songs with the largest and smallest vocabulary in the 1979 album
print(f"The songs with the largest and smallest vocabulary in the 1979 album are: {largestAndSmallestVocabulary(normalized_1979)}\n")
time.sleep(1)
#Introducing intentional waiting for UX purposes
print('Calculating statistics for the 2017 album...')
time.sleep(1)
print('Doing some proper math here...\n')
time.sleep(2)
#Checking the most frequent words in the 2017 album
mostFrequentWordsDict2017 = dict(mostFrequentWords(lyrics_2017))
print(f"The most frequent words in the 2017 album are: {mostFrequentWordsDict2017}\n")
time.sleep(1)
#Checking the type to token ratio in the 2017 album
print(f"The type to token ratio in the 2017 album is: {typeToTokenRatio(lyrics_2017)}\n")
time.sleep(1)
#Checking the average word length in the 2017 album
print(f"The average word length in the 2017 album is: {averageWordLength(lyrics_2017)} words.\n")
time.sleep(1)
#Checking the longest and shortest songs in the 2017 album
print(f"The longest and shortest songs in the 2017 album are: {longestAndShortestSongs(normalized_2017)}\n")
time.sleep(1)
#Checking the songs with the largest and smallest vocabulary in the 2017 album
print(f"The songs with the largest and smallest vocabulary in the 2017 album are: {largestAndSmallestVocabulary(normalized_2017)}\n")
time.sleep(1)

Calculating statistics for the 1979 album...
Conjuring up the statistics...

The most frequent words in the 1979 album are: {'my': 29, 'i': 25, 'you': 21, 'are': 20, 'we': 18, 'and': 16, 'conversation': 16, 'to': 13, 'all': 12, 'the': 11, 'in': 10, 'were': 9, 'is': 9, 'me': 9, 'not': 8}

The type to token ratio in the 1979 album is: 0.36964980544747084

The average word length in the 1979 album is: 3.782101167315175 words.

The songs with the largest and smallest vocabulary in the 1979 album are: ['Metal', 'M.E.']

Calculating statistics for the 2017 album...
Doing some proper math here...

The most frequent words in the 2017 album are: {'i': 67, 'you': 65, 'the': 56, 'when': 44, 'and': 41, 'is': 37, 'a': 35, 'my': 32, 'me': 27, 'for': 23, 'would': 21, 'everything': 20, 'your': 18, 'it': 18, 'we': 16}

The type to token ratio in the 2017 album is: 0.23956043956043957

The average word length in the 2017 album is: 3.7377289377289378 words.

The longest and shortest songs in the 2017 alb

In [7]:
#Defining function that returns the total number of lyrics in a list of lyrics
def totalNumberOfLyrics(lyrics):
    '''This function returns the total number of lyrics in a list of lyrics.
    Input: lyrics (list)
    Output: total number of lyrics (int)'''
    return sum(len(lyric.split()) for lyric in lyrics)
print (f"The total number of lyrics in the 1979 album is: {totalNumberOfLyrics(lyrics_1979)}\n")
time.sleep(1)
print (f"The total number of lyrics in the 2017 album is: {totalNumberOfLyrics(lyrics_2017)}\n")
time.sleep(1)

The total number of lyrics in the 1979 album is: 514

The total number of lyrics in the 2017 album is: 1365



According to the statistics Numan was more verbose (had more lyrics in total) in 1365 lyrics in 1979 than in 2017 with 1005 lyrics.

In addition, the average word length in 1979 was 4.5 words, while in 2017 it was 4.3 words.

I would define repetitiveness by the number of the most repeated top 3 lyrics. In the 19


In [5]:
#Repeating the above steps for solely the song 'Cars' in the 1979 album
#Making a list of the lyrics of the song 'Cars' in the 1979 album
lyrics_1979_cars = [lyrics_1979[0]] #The lyrics of the song 'Cars' are the first element in the list of lyrics of the 1979 album
#Making a list of the normalized lyrics of the song 'Cars' in the 1979 album
normalized_1979_cars = [normalized_1979['Cars']]
#Defining a patched function for longest and shortest songs that also takes song as input
def patched_longestAndShortestSongs(normalized_dictionary, song):
    '''This function returns the longest and shortest songs in a list of lyrics.
    Input: normalized_dictionary (dictionary)
    Output: longest and shortest songs (list)
    '''
    max_song_len = 0
    min_song_len = 9999999
    max_song_name = ''
    min_song_name = ''
    for song in normalized_dictionary:
        l = len(normalized_dictionary[song])
        if l > max_song_len:
            max_song_len = l
            max_song_name = song
        if l < min_song_len:
            min_song_len = l
            min_song_name = song
    return [max_song_name, min_song_name]
#print (normalized_1979_cars)
#Checking the most frequent words in the song 'Cars' in the 1979 album
print('Calculating statistics for the song "Cars" in the 1979 album...\n')
time.sleep(2)
mostFrequentWordsDict1979_cars = dict(mostFrequentWords(lyrics_1979_cars))
print(f"The most frequent words in the song 'Cars' in the 1979 album are the following: {mostFrequentWordsDict1979_cars}\n")
time.sleep(1)
#Checking the type to token ratio in the song 'Cars' in the 1979 album
print(f"The type to token ratio in the song 'Cars' in the 1979 album is: {typeToTokenRatio(lyrics_1979_cars)}\n")
time.sleep(1)
#Checking the average word length in the song 'Cars' in the 1979 album
print(f"The average word length in the song 'Cars' in the 1979 album is: {averageWordLength(lyrics_1979_cars)} words.\n")



Calculating statistics for the song "Cars" in the 1979 album...

The most frequent words in the song 'Cars' in the 1979 album are the following: {'i': 9, 'you': 8, 'the': 7, 'like': 5, 'and': 4, 'to': 4, 'im': 3, 'is': 3, 'could': 3, 'me': 3, 'in': 2, 'they': 2, 'make': 2, 'my': 2, 'metal': 2}

The type to token ratio in the song 'Cars' in the 1979 album is: 0.6384615384615384

The average word length in the song 'Cars' in the 1979 album is: 3.6307692307692307 words.



4. For all the lyrics in each album, construct a word cloud, one per album.

    *   You just need to reassign the variable 'frequencies' in the code below to a dictionary of word frequencies (or Counter) counted from an album. Try to use an output from one of the functions you made above.

In [6]:
import matplotlib.pyplot as plt
from wordcloud import WordCloud

wordcloud_1979 = WordCloud()
wordcloud_1979.generate_from_frequencies(frequencies=mostFrequentWordsDict1979)
plt.figure()
plt.imshow(wordcloud_1979, interpolation="bilinear")
plt.axis("off")
plt.show()

wordcloud_2017 = WordCloud()
wordcloud_2017.generate_from_frequencies(frequencies=mostFrequentWordsDict2017)
plt.figure()
plt.imshow(wordcloud_2017, interpolation="bilinear")
plt.axis("off")
plt.show()


ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.10 from "/Users/george/Desktop/Uni/Development/georges-2023-coding-the-humanities/venv/bin/python"
  * The NumPy version is: "1.24.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'
